### PR TITLE
WIP: Fixes #22823 - support for large puppet environments

### PR DESCRIPTION
**Pull request by dmitri-d**
Please also see the corresponding PR in smart-proxy git repository: https://github.com/theforeman/smart-proxy/pull/570

### CODE REVIEW

File: lib/proxy_api/puppet.rb [hunk 1]
```diff
@@ -29,8 +35,7 @@ def classes(env)
 
     def class_count(env)
       return if env.blank?
-      pcs = parse(get("environments/#{env}/classes"))
-      pcs.length
+      parse(get("environments/#{env}/classes_counter"))
     rescue => e
       raise ProxyException.new(url, e, N_("Unable to get classes from Puppet for %s"), env)
```
Comment (ekohl) [inline lib/proxy_api/puppet.rb:40]: I'd change this to ```get class count``` since it hits a different API endpoint now.

**Review by ekohl — COMMENTED** [review]
It's good to know that this breaks with older proxies. I don't mind this per se, but it should be part of the release notes.

It might also be nicer to show a question mark in the UI or similar rather than 0.

**Comment by dmitri-d** [comment]
* renamed ```ProxyAPI::Puppet#class count``` to ```ProxyAPI::Puppet#get_class count```
* View shows '?' for environments missing class counts instead of '0' (although this should never happen)

**Comment by ekohl** [comment]
> View shows '?' for environments missing class counts instead of '0' (although this should never happen)

I'd think this can happen if you use an older proxy than foreman or a timeout occurs.

**Comment by dmitri-d** [comment]
> I'd think this can happen if you use an older proxy than foreman or a timeout occurs.

An older proxy will return a 404 since this is a new api end-point. Timeout cannot occur -- class counters are updated asynchronously. If puppet environments page is opened shortly after smart-proxy has been (re)started, a 503 error will be returned.

**Comment by ekohl** [comment]
> Timeout cannot occur -- class counters are updated asynchronously.

There's still a network between the foreman and a remote proxy. I meant a HTTP timeout there, not inside the proxy itself.

**Comment by jhutar** [comment]
Hello @witlessbird . By applying these two PRs (https://github.com/theforeman/smart-proxy/pull/570 and https://github.com/theforeman/foreman/pull/5338) to 6.4.0 snap5, I see this error in the production.log:

```
2018-06-07T08:31:53 [W|dyn|] Error on on_execution_plan_save event
NoMethodError: undefined method `caller_execution_plan_id' for #<SendExpireSoonNotifications:0x00000000095bfb40>
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-0.13.1/app/models/foreman_tasks/task/dynflow_task.rb:18:in `update_from_dynflow'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-0.13.1/lib/foreman_tasks/dynflow/persistence.rb:43:in `on_execution_plan_save'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-0.13.1/lib/foreman_tasks/dynflow/persistence.rb:12:in `block in save_execution_plan'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-0.13.1/lib/foreman_tasks/dynflow/persistence.rb:10:in `tap'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman-tasks-0.13.1/lib/foreman_tasks/dynflow/persistence.rb:10:in `save_execution_plan'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/persistence.rb:73:in `save_execution_plan'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/execution_plan.rb:429:in `save'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/execution_plan.rb:132:in `update_state'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/execution_plan.rb:290:in `plan'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/delayed_plan.rb:24:in `plan'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/delayed_executors/abstract_core.rb:60:in `block (3 levels) in process'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/executors.rb:12:in `run_user_code'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/delayed_executors/abstract_core.rb:59:in `block (2 levels) in process'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/delayed_executors/abstract_core.rb:39:in `with_error_handling'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/delayed_executors/abstract_core.rb:50:in `block in process'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/delayed_executors/abstract_core.rb:48:in `each'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/delayed_executors/abstract_core.rb:48:in `process'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/delayed_executors/polling.rb:26:in `check_delayed_plans'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/delayed_executors/polling.rb:20:in `start'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/actor.rb:6:in `on_message'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/context.rb:46:in `on_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/executes_context.rb:7:in `on_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.0.3/lib/dynflow/actor.rb:26:in `on_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/awaits.rb:15:in `on_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/sets_results.rb:14:in `on_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/buffer.rb:38:in `process_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/buffer.rb:31:in `process_envelopes?'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/buffer.rb:20:in `on_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/termination.rb:55:in `on_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/removes_child.rb:10:in `on_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/abstract.rb:25:in `pass'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/behaviour/sets_results.rb:14:in `on_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/core.rb:161:in `process_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/core.rb:95:in `block in on_envelope'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/core.rb:118:in `block (2 levels) in schedule_execution'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/synchronization/mri_lockable_object.rb:38:in `block in synchronize'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/synchronization/mri_lockable_object.rb:38:in `synchronize'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/synchronization/mri_lockable_object.rb:38:in `synchronize'
/opt/theforeman/tfm/root/usr/share/gems/gems/concurrent-ruby-edge-0.2.4/lib/concurrent/actor/core.rb:115:in `block in schedule_execution'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/serialized_execution.rb:18:in `call'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/serialized_execution.rb:96:in `work'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/serialized_execution.rb:77:in `block in call_job'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:348:in `run_task'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:337:in `block (3 levels) in create_worker'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `loop'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `block (2 levels) in create_worker'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `catch'
/opt/theforeman/tfm-ror51/root/usr/share/gems/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `block in create_worker'
/opt/theforeman/tfm/root/usr/share/gems/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
2018-06-07T08:31:53 [W|dyn|] Error on on_execution_plan_save event
NoMethodError: undefined method `caller_execution_plan_id' for #<SendExpireSoonNotifications:0x00000000098a4720>
```

**Comment by ares** [comment]
@ewoud we know this is still the a common problem, do you have any thoughts on the WIP solution here? Should we try to resume and build on top of this or is there other solution that should be investigated to mitigate the issue? Please let us know before next triage (2 weeks from now), thank you!

**Comment by ares** [comment]
Pinging the right person now - @ekohl , please take a look.

**Comment by ekohl** [comment]
I'm closing this. On the smart proxy side this is already closed. Because we can rely on Puppetserver being fast (where with Puppet 3 we parsed ourselves), this is much less of an issue nowadays. However, there's still some quirks around this.

**Pull request closed without merging**